opentimelineio_contrib/adapters/xges.py

```diff
@@ -0,0 +1,774 @@
+#
+# You may obtain a copy of the Apache License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the Apache License with the above modification is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied. See the Apache License for the specific
+# language governing permissions and limitations under the Apache License.
+#
+
+"""OpenTimelineIO GStreamer Editing Services XML Adapter. """
+import os
+import re
+import unittest
+
+from collections import defaultdict
+from decimal import Decimal
+from fractions import Fraction
+from xml.etree import cElementTree
+from xml.dom import minidom
+import opentimelineio as otio
+
+META_NAMESPACE = "XGES"
+
+
+FRAMERATE_FRAMEDURATION = {23.98: "24000/1001",
+                           24: "600/25",
+                           25: "25/1",
+                           29.97: "30000/1001",
+                           30: "30/1",
+                           50: "50/1",
+                           59.94: "60000/1001",
+                           60: "60/1"}
+
+
+TRANSITION_MAP = {
+    "crossfade": otio.schema.TransitionTypes.SMPTE_Dissolve
+}
+# Two way map
+TRANSITION_MAP.update(dict([(v, k) for k, v in TRANSITION_MAP.items()]))
+
+
+class GstStructure(object):
+    """
+    GstStructure parser with a "dictionnary" like API.
+    """
+    UNESCAPE = re.compile(r'(?<!\\)\\(.)')
+    INT_TYPES = "".join(
+        ("int", "uint", "int8", "uint8", "int16",
+         "uint16", "int32", "uint32", "int64", "uint64")
+    )
+
+    def __init__(self, text):
+        self.text = text
+        self.modified = False
+        self.name, self.types, self.values = GstStructure._parse(text + ";")
+
+    def __repr__(self):
+        if not self.modified:
+            return self.text
+
+        res = self.name
+        for key, value in self.values.items():
+            value_type = self.types[key]
+            res += ', %s=(%s)"%s"' % (key, value_type, self.escape(value))
+        res += ';'
+
+        return res
+
+    def __getitem__(self, key):
+        return self.values[key]
+
+    def set(self, key, value_type, value):
+        if self.types.get(key) == value_type and self.values.get(key) == value:
+            return
+
+        self.modified = True
+        self.types[key] = value_type
+        self.values[key] = value
+
+    def get(self, key, default=None):
+        return self.values.get(key, default)
+
+    @staticmethod
+    def _find_eos(s):
+        # find next '"' without preceeding '\'
+        l = 0
+        while 1:  # faster than regexp for '[^\\]\"'
+            p = s.index('"')
+            l += p + 1
+            if s[p - 1] != '\\':
+                return l
+            s = s[(p + 1):]
+        return -1
+
+    @staticmethod
+    def escape(s):
+        if not isinstance(s, str):
```

Comment:
Should unicode be supported also?

Reply:
Done with the same trick as in the aaf adapter to handle py2 and 3 in the same code path.